src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +28,69 @@ public function getUrl(string $bucket, string $key): string
         return $this->getEndpoint($uri, [], null);
     }
 
-    /*
-     * TODO Add get() and put() which supports large files
-     * TODO add has()
+    public function download(string $bucket, string $key): ResultStream
+    {
+        return $this->getObject(['Bucket' => $bucket, 'Key' => $key])->getBody();
+    }
+
+    public function has(string $bucket, string $key): bool
+    {
+        return $this->objectExists(['Bucket' => $bucket, 'Key' => $key])->isSuccess();
+    }
+
+    /**
+     * @param string|resource|callable|iterable $object
      */
+    public function upload(string $bucket, string $key, $object): void
+    {
+        // Send upload requests with size 64 MB
+        $chunkSize = 64 * 1024 * 1024;
+        $directUpload = \is_string($object);
+        if (!$directUpload && \is_resource($object)) {
+            // Try to find size of the resource
+            $stats = \fstat($object);
+            if (isset($stats['size']) && $stats['size'] < $chunkSize) {
+                $directUpload = true;
+            }
+        }
+
+        if ($directUpload) {
+            $this->putObject(['Bucket' => $bucket, 'Key' => $key, 'Body' => $object]);
+
+            return;
+        }
+
+        if (\is_resource($object)) {
+            $stream = ResourceStream::create($object, $chunkSize);
+        } elseif (\is_callable($object)) {
+            $stream = CallableStream::create($object, $chunkSize);
+        } elseif (\is_iterable($object)) {
+            $stream = IterableStream::create($object);
+        } else {
+            throw new InvalidArgument('Third argument of SimpleS3Client::upload() must be string, resource, callable or an interable.');
+        }
+
+        $uploadId = $this->createMultipartUpload(['Bucket' => $bucket, 'Key' => $key])->getUploadId();
```

Comment:
should we delete this if upload failed on a part? or is it handled by AWS ?

Reply:
Maybe. I'll check

Reply:
Im now catching any errors and aborting the upload.